pkg/plugins/leader/postgres/leader_elector.go

```diff
@@ -0,0 +1,93 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"cirello.io/pglock"
+
+	"github.com/Kong/kuma/pkg/core"
+	"github.com/Kong/kuma/pkg/core/runtime/component"
+)
+
+var log = core.Log.WithName("postgres-leader")
+
+const kumaLockName = "kuma-cp-lock"
+
+// postgresLeaderElector implements leader election using PostgreSQL DB.
+// pglock does not rely on timestamps, which eliminates the problem of clock skews, but the cost is that first leader election can happen only after lease duration
+// pglock does optimistic locking under the hood, the alternative would be to use pg_advisory_lock
+type postgresLeaderElector struct {
+	leader     int32
+	lockClient *pglock.Client
+	callbacks  []component.LeaderCallbacks
+}
+
+var _ component.LeaderElector = &postgresLeaderElector{}
+
+func NewPostgresLeaderElector(lockClient *pglock.Client) component.LeaderElector {
+	return &postgresLeaderElector{
+		lockClient: lockClient,
+	}
+}
+
+func (p *postgresLeaderElector) Start(stop <-chan struct{}) {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	go func() {
+		<-stop
+		cancelFn()
+	}()
+
+	for {
```

Comment:
Added check if stop channel is closed. Additionaly I added 3s gracefully period when CP is stopping so all components have a chance to be stopped. Otherwise some components managed to be closed gracefully, some didn't.